### ISSUE TITLE
Conditional install for UseDotNetV2 ?

### ISSUE THREAD

**Issue by hiyadav** [microsoft/azure-pipelines-tasks#10405]
Copied issue from vsts-docs=> https://github.com/MicrosoftDocs/vsts-docs/issues/4153 

If we specify a version that is already installed, will this step recognize that and skip reinstalling, or do we need a guard script?

**Comment by hiyadav** [comment]
Possible solutions:

1. Update azure-pipelines-image-generation to create the markup file for each version installed on agent image. 
2. Provide a checkbox for not installing the version if already present on the machine.
3. Provide a path to search for if version is installed at that location. (like /home/dotnet/)

Complications with each:
1. Users will have to global install locations as the installation path. This might lead to problems on the agents.
2. The global install location in linux machines is not set.
2. & 3. Enable Multi-Level lookup needs be ignore if any of the two cases are being used.

**Comment by MarkKharitonov** [comment]
Is there an ETA for this? We have an issue where cisco has blacklisted the IP address 199.232.36.133 and all of our pipelines are failing now because of this. If this issue was fixed, at least we could have fetched it from the cache until the cisco issue is resolved.

Here is what we get:
Failed to download or parse release-index.json with error: {"errno":"ETIMEDOUT","code":"ETIMEDOUT","syscall":"connect","address":"199.232.36.133","port":443}

**Comment by StefanOssendorf** [comment]
Even the multi level lookup as documented here [Dotnet core installer doc](https://github.com/MicrosoftDocs/azure-devops-docs/blob/master/docs/pipelines/tasks/tool/dotnet-core-tool-installer.md) does not work as described. We are using the Windows-2019 hosted agent where according to this [Hosted agent installed software readme](https://github.com/actions/virtual-environments/blob/master/images/win/Windows2019-Readme.md) the used runtime should already be installed and not be re-installed.

**Comment by MarkKharitonov** [comment]
Any update on this?

**Comment by mattgenious** [comment]
This seems like it would remove a lot of unnecessary queue time in pipelines and download traffic from sdk sources if applied.
What is the progress on this?

**Comment by StefanOssendorf** [comment]
Any news on that?

**Comment by rmandvikar** [comment]
Any eta on this please? For me every pipeline run takes 30s to download any sdk which is waste of time.

**Comment by StefanOssendorf** [comment]
Not only a waste of time. It also uses maybe precious seconds from your free time per month.

**Comment by pulkitaggarwl** [comment]
@aaronhallberg can you have someone have a look at this ?

**Comment by StefanOssendorf** [comment]
Still nothing, huh?

**Comment by rmandvikar** [comment]
Just saying. I'm also eagerly waiting for this fix. The pipelines are such a timesink currently. It takes more time to download the sdk (30s) than it does to run all the other steps combined (<30s). An update is really appreciated.

**Comment by LennardF1989** [comment]
How can this still be open after all these years? Downloading .NET 5.x takes 30 seconds while it's already on the windows-latest image. This costs us time, and it costs Microsoft precious server resources.

**Comment by mattgenious** [comment]
> How can this still be open after all these years? Downloading .NET 5.x takes 30 seconds while it's already on the windows-latest image. This costs us time, and it costs Microsoft precious server resources. It works fine for the NodeTool task, why can't this just work as well?

Agreed, number of builds times 30 seconds should be an absolutely insane amount of server capacity freed up if this is fixed.

**Comment by bbieniek** [comment]
I have spent the entire day today trying to figure this out.
It's pretty interesting.

When you create a new pipeline and use the task:

`- task: UseDotNet@2
    displayName: 'Install .NET'
    inputs:
      version: '6.0.x'
      includePreviewVersions: true`

Then it will not find anything in cache.
When you rerun it, it will not find anything in cache.
It will however install .NET in the version **6.0.301**

However, when I ran a simple script to display the dotnet version, it turns out that the dotnet version is **6.0.301**.

So, to summarize - UseDotnet doesn't see that the .NET 6 is installed by default.

This bug has been opened for 2 years and each .NET install consumes 250MB and 30s of time.
I wonder how many builds have been executed with this command...

I wrote a simple workaround to cache the .NET installation.

`# Starter pipeline
# Start with a minimal pipeline that you can customize to build and deploy your code.
# Add steps that build, run tests, deploy, and more:
# https://aka.ms/yaml

variables:
  DOTNET_CACHE_FOLDER: $(System.DefaultWorkingDirectory)/dotnet
  INSTALL_DOTNET: true
  dotnet: '6.0.x'

trigger:
- master

pool:
  vmImage: ubuntu-latest

jobs:
- job: 'test'
  steps:
  - task: PowerShell@2
    displayName: Get dotnet version
    inputs: 
      targetType: inline
      script: |
        try {
          $version = (dotnet --version)
          Write-Host "Found version: $version"        
          $expectedVersion = '$(dotnet)'
          Write-Host "Expected version: $expectedVersion"
          if ($version) {        
            if ($version -like $expectedVersion.Replace('x', '*')) 
            {
                    Write-Host "The installed version $version matches the expected dotnet version $expectedVersion"
                    Write-Host "##vso[task.setVariable variable=INSTALL_DOTNET;]$false"
            }
            else {
              Write-Host "The installed version $version doesnt match the expected dotnet version $expectedVersion"
            }
          }
          else {
            Write-Host "Dotnet not found"
          }        
        }
        catch {
          Write-Host "Dotnet not found"
        }
    
  - task: Cache@2
    inputs:
      key: '"dotnet" | "$(Agent.OS)" | "$(dotnet)"'
      path: $(Agent.ToolsDirectory)/dotnet
      cacheHitVar: CACHE_RESTORED
    displayName: Cache Dotnet packages
    condition: eq(variables['INSTALL_DOTNET'], 'True')
    
  - task: UseDotNet@2
    displayName: 'Install .NET'
    inputs:
      version: $(dotnet)
      includePreviewVersions: true
      installationPath: $(Agent.ToolsDirectory)/dotnet
    condition: eq(variables['INSTALL_DOTNET'], 'True')
`

**Comment by rmandvikar** [comment]
@pulkitaggarwl Any update on this please? Quite a few people, including me, are waiting on this. It's confusing in the current state, as either the doc should be fixed, or the bug acknowledged and fixed. I read [this thread](https://developercommunity.visualstudio.com/t/use-net-core-sdk-task-for-3x-reports-version-31410/1468875), and can't tell if this is a bug or expected behavior. Please clarify.

**Comment by KerbalSpike** [comment]
It is really mind-boggling, that there has not been any fix for this. I mean I could understand it if it would only be affecting and annoying users, but this must be a pretty big resource hog for Microsoft's infrastructure, assuming that there must be thousands of pipelines running with this task each and every day...

**Comment by weshaggard** [comment]
@bishal-pdMSFT are there any plans to add this support? It is definitely an issue that lots of folks have to deal with and potentially make workarounds for.

**Comment by doggy8088** [comment]
I have a workaround for this issue.

Don't use the `UseDotNet@2` task, just create a `global.json` file before your `dotnet build` by using Command Line (CMD or Bash). 

> Don't use `dotnet new globaljson --sdk-version 6.0.100 --roll-forward latestFeature` for this!

Here is an example:

Command Prompt

```bat
echo {"sdk":{"version":"6.0.100","rollForward":"latestFeature","allowPrerelease":false}} > global.json
```

Bash

```sh
echo '{"sdk":{"version":"6.0.100","rollForward":"latestFeature","allowPrerelease":false}}' > global.json
```

**Comment by npbenjohnson** [comment]
Setting a global.json works if the version you want is definitely installed, but it will break when not installed, this workaround fixes the task's cacheing:
There is a workaround here: https://github.com/actions/runner-images/issues/4884#issuecomment-1013047819

**Comment by IvanAlekseev** [comment]
It is still the same for us.
Agent info:
```
Operating System
Ubuntu
22.04.3
LTS
Runner Image
Image: ubuntu-22.04
Version: 20240107.1.0
Included Software: https://github.com/actions/runner-images/blob/ubuntu22/20240107.1/images/ubuntu/Ubuntu2204-Readme.md
Image Release: https://github.com/actions/runner-images/releases/tag/ubuntu22%2F20240107.1
Runner Image Provisioner
2.0.321.1
Current image version: '20240107.1.0'
```

Task
```yaml
          - task: UseDotNet@2
            inputs:
              packageType: sdk
              version: 8.0.x
              displayName: dotnet install
```

Logs
```
##[section]Starting: UseDotNet
==============================================================================
Task         : Use .NET Core
Description  : Acquires a specific version of the .NET Core SDK from the internet or the local cache and adds it to the PATH. Use this task to change the version of .NET Core used in subsequent tasks. Additionally provides proxy support.
Version      : 2.232.0
Author       : Microsoft Corporation
Help         : https://aka.ms/AA4xgy0
==============================================================================
Tool to install: .NET Core sdk version 8.0.x.
Found version 8.0.100 in channel 8.0 for user specified version spec: 8.0.x
get-os-distro: Error: Distribution specific OS name and version could not be detected: UName = Linux
Version 8.0.100 was not found in cache.
Getting URL to download .NET Core sdk version: 8.0.100.
Detecting OS platform to find correct download package for the OS.
[command]/home/vsts/work/_tasks/UseDotNet_b0ce7256-7898-45d3-9cb5-176b752bfea6/2.232.0/externals/get-os-distro.sh
Primary:linux-x64
Legacy:-x64
Detected platform (Primary): linux-x64
Detected platform (Legacy): -x64
Downloading: https://download.visualstudio.microsoft.com/download/pr/5226a5fa-8c0b-474f-b79a-8984ad7c5beb/3113ccbf789c9fd29972835f0f334b7a/dotnet-sdk-8.0.100-linux-x64.tar.gz
Extracting downloaded package /home/vsts/work/_temp/7e80ed41-c353-4e64-873e-f580fa4c8360.
Extracting archive
[command]/usr/bin/tar xC /home/vsts/work/_temp/cea2d5f9-0ff5-46c5-86b6-7da4a5a77cd3 -f /home/vsts/work/_temp/7e80ed41-c353-4e64-873e-f580fa4c8360
Successfully installed .NET Core sdk version 8.0.100.
Creating global tool path and pre-pending to PATH.
##[section]Finishing: UseDotNet
```

**Comment by mattgenious** [comment]
Oh right, I just deleted the install step but then obviously that introduces the uncertainty of not guaranteeing the sdk version being installed, makes sense.

**Comment by StefanOssendorf** [comment]
Still nothing? This is now nearly 5 years old and could have saved a lot of time and resources...

**Comment by StefanOssendorf** [comment]
Still nothing? Wasting time, space and energy...

**Comment by tillig** [comment]
Still an issue, last I checked.

**Comment by Frulfump** [comment]
Yeah this is getting ridiculous, it's labeled q3 since October 29:th 2019 and I guess they didn't specify which year 😅.
Please could any of the assignees comment @bishal-pdMSFT @hiyadav @pulkitaggarwl there's even a PR for you to review that fixes that issue but it just sits without attention https://github.com/microsoft/azure-pipelines-tasks/pull/19415
cc @DergachevE

**Comment by DergachevE** [comment]
Tagging @tarunramsinghani

**Comment by Frulfump** [comment]
@tarunramsinghani There's a lot of dupe issues for this please clean up the issues and look at the PR #19415

**Comment by Frulfump** [comment]
Still an issue, we are waiting for the maintainers to take action.

**Comment by CaringDev** [comment]
Does `checkForExistingVersion` resolve this?

**Comment by Frulfump** [comment]
> Does `checkForExistingVersion` resolve this?

Yes presumably, thanks

Checked the docs
https://learn.microsoft.com/en-us/azure/devops/pipelines/tasks/reference/use-dotnet-v2?view=azure-pipelines
checkForExistingVersion - Check for existing installation
boolean. Default value: false.

> Select if you want to detect if the specified version is already installed before attempting to download. Use only when installationPath is Empty i.e. set to default value.

Why anyone would want this set to false I have no idea but that was the default that was chosen so it remains opt-in to get the sane behavior.
